fedbiomed/common/config.py

```diff
@@ -0,0 +1,210 @@
+
+    def read(self) -> bool:
+        """Reads configuration file that is already existing in given path
+
+        Raises verision compatibility error
+        """
+
+        self._cfg.read(self.path)
+
+        # Validate config version
+        raise_for_version_compatibility(
+            self._cfg["default"]["version"],
+            self.CONFIG_VERSION,
+            f"Configuration file {self.path}: found version %s expected version %s")
+
+        return True
+
+    def get(self, section, key) -> str:
+        """Returns value for given ket and section"""
+
+        return self._cfg.get(section, key)
+
+    def sections(self) -> list:
+        """Returns sections of the config"""
+
+        return self._cfg.sections()
+
+    def write(self):
+        """Writes config file"""
+
+        try:
+            with open(self.path, 'w') as f:
+                self._cfg.write(f)
+        except configparser.Error:
+            raise IOError(ErrorNumbers.FB600.value + ": cannot save config file: " + self.path)
+
+
+    def generate(self, force: bool = False) -> bool:
+        """"Generate configuration file
+
+        Args:
+        force: Overwrites existing configration file
+        """
+
+        # Check if configuration is already existing
+        if self.is_config_existing() and not force:
+            return self.read()
+
+        component_id = f"{self.COMPONENT_TYPE}_{uuid.uuid4()}"
+
+        self._cfg['default'] = {
+            'id': component_id,
+            'component': self.COMPONENT_TYPE,
+            'version': str(self.CONFIG_VERSION)
+        }
+
+        # DB PATH RELATIVE
+        db_path  = os.path.join(self.root, VAR_FOLDER_NAME, f"{DB_PREFIX}{component_id}.json")
+        self._cfg['default']['db'] = os.path.relpath(db_path, os.path.join(self.root, CONFIG_FOLDER_NAME))
+
+        ip, port = retrieve_ip_and_port(self.root)
+        allow_default_biprimes = os.getenv('ALLOW_DEFAULT_BIPRIMES', True)
+
+        # Generate self-signed certificates
+        key_file, pem_file = generate_certificate(
+            root=self.root, 
+            component_id=component_id, 
+            prefix=MPSPDZ_certificate_prefix)
+
+        self._cfg['mpspdz'] = {
+            'private_key': os.path.relpath(key_file, os.path.join(self.root, 'etc')),
+            'public_key': os.path.relpath(pem_file, os.path.join(self.root, 'etc')),
+            'mpspdz_ip': ip,
+            'mpspdz_port': port,
+            'allow_default_biprimes': allow_default_biprimes,
+            'default_biprimes_dir': os.path.relpath(
+                os.path.join(self.root, 'envs', 'common', 'default_biprimes'),
+                os.path.join(self.root, 'etc')
+            )
+        }
+
+        # Calls child class add_parameterss
+        self.add_parameters()
+
+        # Write configuration file
+        return self.write()
+
+    @abstractmethod
+    def add_parameters(self):
+        """"Component specific argument creation"""
+
+
+class NodeConfig(Config):
```

Comment:
Hi @mvesin, 

I wanted to implemented it as independently in `common.cli` .  If I would add it `researhcer.config` `common.cli` would have to import `researcher` and `node` modules.

Reply:
Hi @mvesin, 
Thanks, I will check your ideas and see how it will go.

Reply:
Hi @mvesin, 
I tried to implement your suggestions. However, it does not behave as it is expected. Since researcher cli import researcher environ, and researcher environ instantiates `ResearcherConfig` it automatically create a confgiuration file if there is no existing.  Therefore, the cli has no control over the configuration creation. So now I remember why I wanted to implement these config classes in the common module and intialize argument parser in `common.cli`. 

I don't know what do you think but I would prefer controlling configuration creation independently from environ class, and I couldn't find a way to do it smoothly from the component specific cli module since they all import environ class which depends on configuration file.

Reply:
What I would suggest is to conditionally and dynamically import `ResearcherConfig` and `NodeConfig` in `common.cli``

Here is the `_create_component` method of `common.cli`. 
```python
 if args.component.lower() == "node":
            NodeConfig = importlib.import_modue('fedbiomed.node.config').NodeConfig
            config = NodeConfig()
        elif args.component.lower() == "researcher":
            ResearcherConfig = importlib.import_modue('fedbiomed.researcher.config').ResearcherConfig
            config = ResearcherConfig(root=args.root, name=args.name, auto_generate=False)
        else: 
            print(f"Undefined component type {args.component}")
            return
```

You may name it as dirty solution :) but i think it is quite important to not to mix environ and config classes. What I wanted to do in this implementation was to handle config parsing and verification before starting the application/node/researcher. Than it is environ class that is responsible of reading config file and overwriting some value based on user decision.

Reply:
Thank you Sergen, I missed that point.
Nevertheless doesn't the same issue exist with current implementation (automatic creation of configuration) ?

```
source ./scripts/fedbiomed_environment clean
source ./scripts/fedbiomed_environment researcher
python -m fedbiomed.node.cli --list
```

... and we have a node config file created (but maybe it is linked to your last comment here https://github.com/fedbiomed/fedbiomed/pull/919#discussion_r1415558000

Reply:
Yes because of `environ` class it is automatically created as a short cut to lunch components quickly. It is not a problem but when launching node. It will be problem when running configuration create `fedbiomed_run node configuration create --name test.ini' because there CLI 
- first: imports `environ` that instantiate `NodeConfig`  that creates configuration file or reads existing one. 
- second: it will go in arg parser for `_create_component` and try to recreate configuration however it will complain because there is already one created. 
- or there will be problem of generating two distinct configuration. 

Therefore, I wanted to separate Config management and Environ management, along with the new positional argument such as `--force` to overwrite config file. This approach will be much more flexible if once bashc CLI and python CLI redesigned to be merged in single Python CLI.

If you think that it is not necessary I can revert back to the version in the develop.

Reply:
I pushed a new version that implements the suggestion in this comment https://github.com/fedbiomed/fedbiomed/pull/919#discussion_r1415579178